Buy/Graph.Task.swift

```diff
@@ -26,42 +26,271 @@
 
 import Foundation
 
-public extension Graph {
+/// Provides an interface for managing the `URLSessionDataTask` created by `Graph.Client`. Since the
+/// the underlying `URLSessionDataTask` can change (in the case of retried requests), `Task` provides
+/// a single place to `cancel()` or `resume()` this underlying task.
+///
+public protocol Task {
     
-    /// Provides an interface for managing the `URLSessionDataTask` created by `Graph.Client`. Since the 
-    /// the underlying `URLSessionDataTask` can change (in the case of retried requests), `Task` provides
-    /// a single place to `cancel()` or `resume()` this underlying task.
-    ///
-    public class Task {
+    /// Starts the underlying task
+    func resume()
+    
+    /// Cancel the underlying task
+    func cancel()
+}
+
+internal extension Graph {
+    
+    internal class InternalTask<R: GraphQL.AbstractResponse>: Task {
+        
+        typealias TaskCompletion = (R?, QueryError?) -> Void
         
-        internal private(set) var task: URLSessionDataTask
+        internal private(set) var isResumed:   Bool = false
+        internal private(set) var isCancelled: Bool = false
+        
+        internal let cache:        Cache
+        internal let session:      URLSession
+        internal let request:      URLRequest
+        internal let cachePolicy:  CachePolicy
+        internal let retryHandler: RetryHandler<R>?
+        internal let completion:   TaskCompletion
+        
+        internal var task:         URLSessionDataTask?
         
         // ----------------------------------
         //  MARK: - Init -
         //
-        internal init(representing task: URLSessionDataTask) {
-            self.task = task
+        internal init(session: URLSession, cache: Cache, request: URLRequest, cachePolicy:  CachePolicy, retryHandler: RetryHandler<R>? = nil, completion: @escaping TaskCompletion) {
+            self.cache        = cache
+            self.session      = session
+            self.request      = request
+            self.cachePolicy  = retryHandler == nil ? cachePolicy : .networkOnly
+            self.retryHandler = retryHandler
+            self.completion   = completion
         }
         
         // ----------------------------------
-        //  MARK: - Setters -
+        //  MARK: - Control -
         //
-        internal func setTask(_ task: URLSessionDataTask) {
-            self.task = task
+        func resume() {
+            self.isResumed = true
+            self.resume(using: self.cachePolicy)
+        }
+        
+        private func resume(using cachePolicy: CachePolicy) {
+            
+            let request    = self.request
+            let completion = self.completion
+            let hash       = request.hash
+            
+            switch cachePolicy {
+            case .cacheFirst(let expireIn):
```

Comment:
I'm not sure I see the benefit of making the `CachePolicy` any smarter than it currently is. Managing the request flow isn't the responsibility of the `CachePolicy`. It shouldn't know anything beyond simple enum values. Otherwise it'll need to know about the `client`, `cache`, `retryHandler`, etc.